homeassistant/helpers/entity_platform.py

```diff
@@ -402,11 +402,12 @@ async def async_add_entities(
         ):
             return
 
-        self._async_unsub_polling = async_track_time_interval(
-            self.hass,
-            self._update_entity_states,
-            self.scan_interval,
-        )
+        if self.scan_interval > timedelta(seconds=0):
```

Comment:
testing for scan_interval being positive is a good thing, but would it not be better to do it inside async_track_time_interval() ?

This should also be done in the other time track function.

Reply:
Not sure.
From a pure OOO point-of-view, I prefer to not to create an object rather than an empty/NOP one, but that's me...

Reply:
Why would you create an object ? Since scan_interval <= 0 is a real bug, you could just raise an exception or return None (but that would probably raise more problems).

Yes it touches a lot….but your change touches something like 5-600 integrations I consider that a lot too.

Reply:
Btw since this is a fundamental change, you need to add tests.

Reply:
> Why would you create an object

Ah, nvm. I had in mind `async_track_time_interval` was a class.

Well, same for me. I'd just want a second opinion to avoid back and forth.

Thanks

Reply:
A scan interval of 0 should not be allowed. Instead of silently ignoring it, we should  disallowing it.

I think step 1 is to update the validation of the command line sensor instead.

Reply:
`command_line` is just an example, here. The integration itself doesn't even have a notion of scan interval.
Validator is in the base PLATFORM_SCHEMA (https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/config_validation.py#L854). 
There is a `positive_time_period`, that I guess could replace `time_period` but that one accepts zero as well. Should a `strictly_positive_time_period` be created?

Now, that would fix 1 stone of this PR, but I actually stumbled on the issue trying to disable scan interval altogether on "utility" sensors, to have "scan on demand". 
What's the opinion on that? Is it something not wanted? Should it be done in another way?

Reply:
If an integration has polling entities, you can already disable it:

![image](https://user-images.githubusercontent.com/1444314/134109626-5e04d4d1-7c0f-419a-bf5e-f36cf6ba0278.png)

Reply:
Basically my point:
- Integrations with a `config_flow` can have polling disabled via the GUI
- There is *no* such mechanism for YAML platforms (to the best of my knowledge)
- "Utilities" integrations (command_line, REST, ...) have not been migrated to config_flow (nor do I think they should)
- Even if they get migrated at some point, it doesn't make much sense to disable polling globally for those

Reply:
Any place where we allow configuring scan interval should we mandate at least 5 seconds. That should be the initial fix. If we want to support a "never" keyword, we can do that, but it should be explicit and we should first discuss that in the architecture repo.

Reply:
Updated as a WIP, to confirm approach and scope.
Especially, should this PR update all integrations with a SCAN_INTERVAL and a `cv.time_period` validator?

Reply:
Updating all integrations would be a mistake.

We recently had an interesting discussion and a larger number of issues, because modbus was changed to minimum SCAN_INTERVAL = 5 seconds. The discussion ended with an acceptance that modbus could use a lower SCAN_INTERVAL.

Also you will find a number of integrations that do not use cv.time_period, but cv.positive_integer.

If you go ahead and update all integrations, it is unknown how many problems we will run into and I highly recommend leaving this to the respective codeowners.